### ISSUE TITLE
[BUG] Too many items in ListBox with ScrollBar tanks performance

### ISSUE THREAD

**Issue by max-moser** [urwid/urwid#845]
##### Description:

I have a scrollbar wrapped around a listbox with _lots_ of widgets ([>28745 of them](https://imgflip.com/i/8h8e5r)) which are focusable for navigation.
However, at this scale the rendering of the scrollbar/listbox starts becoming painfully slow (in the order of seconds on my Thinkpad T14), including the re-rendering on each navigational keypress.

##### Affected versions
Latest stable version from `pypi` (2.6.5)

##### Details

With some debugging, I saw that the problem seems to be in `ScrollBar.render()`, where the [size of the listbox is queried](https://github.com/urwid/urwid/blob/master/urwid/widget/scrollable.py#L494).
This causes the listbox to [iterate over all its contents](https://github.com/urwid/urwid/blob/master/urwid/widget/listbox.py#L572-L575) every time because the `size` parameter causes the caching to be avoided.

##### Possible solution (tested)

When replacing the default `urwid.ListBox` with the following implementation, navigation is _much_ smoother (after the first calculation):

```python
class CachingListBox(uw.ListBox):
    """ListBox with caching for sizes."""

    def __init__(self, body):
        """Constructor."""
        super().__init__(body)
        self._cached_sizes = {}

    def rows_max(self, size, focus):
        """Scrollable protocol for sized iterable and not wrapped around contents."""
        if size in self._cached_sizes:
            return self._cached_sizes[size]

        result = super().rows_max(size, focus)
        self._cached_sizes[size] = result
        return result
```

It wasn't possible to set only a single `self._cached_size` value, because it seems like the `rows_max` are queried with and without the scrollbar width taken into account, which basically immediately invalidates the cache.

Because the listbox content is static for us, I haven't added logic for invalidating the cache when its content changes.

If you like, I can start working on a PR for adding such caching logic to the default `ListBox` class.

**Comment by penguinolog** [comment]
IMHO we could cache `rows_max` with cache invalidation via "modified callback" on list walker.
Due to not all list walkers support such `set_modified_callback` API, we should cache only if was able to set such callback.

**Comment by max-moser** [comment]
Closing as it's fixed by #858
